lib/ui/painting/paint.cc

```diff
@@ -331,6 +331,97 @@ bool Paint::sync_to(DisplayListBuilder* builder,
+  if (isNull()) {
+    return;
+  }
+  FML_DCHECK(paint == DlPaint());
+
+  tonic::DartByteData byte_data(paint_data_);
+  FML_CHECK(byte_data.length_in_bytes() == kDataByteCount);
+
+  const uint32_t* uint_data = static_cast<const uint32_t*>(byte_data.data());
+  const float* float_data = static_cast<const float*>(byte_data.data());
+
+  Dart_Handle values[kObjectCount];
+  if (!Dart_IsNull(paint_objects_)) {
+    FML_DCHECK(Dart_IsList(paint_objects_));
+    intptr_t length = 0;
+    Dart_ListLength(paint_objects_, &length);
+
+    FML_CHECK(length == kObjectCount);
+    if (Dart_IsError(
+            Dart_ListGetRange(paint_objects_, 0, kObjectCount, values))) {
+      return;
+    }
+
+    Dart_Handle shader = values[kShaderIndex];
+    if (!Dart_IsNull(shader)) {
+      if (Shader* decoded = tonic::DartConverter<Shader*>::FromDart(shader)) {
+        auto sampling =
+            ImageFilter::SamplingFromIndex(uint_data[kFilterQualityIndex]);
+        paint.setColorSource(decoded->shader(sampling).get());
+      }
+    }
+
+    Dart_Handle color_filter = values[kColorFilterIndex];
+    if (!Dart_IsNull(color_filter)) {
+      ColorFilter* decoded =
+          tonic::DartConverter<ColorFilter*>::FromDart(color_filter);
+      paint.setColorFilter(decoded->dl_filter());
+    }
+
+    Dart_Handle image_filter = values[kImageFilterIndex];
+    if (!Dart_IsNull(image_filter)) {
+      ImageFilter* decoded =
+          tonic::DartConverter<ImageFilter*>::FromDart(image_filter);
+      paint.setImageFilter(decoded->dl_filter());
+    }
+  }
+
+  paint.setAntiAlias(uint_data[kIsAntiAliasIndex] == 0);
+
+  uint32_t encoded_color = uint_data[kColorIndex];
+  paint.setColor(encoded_color ^ kColorDefault);
+
+  uint32_t encoded_blend_mode = uint_data[kBlendModeIndex];
+  uint32_t blend_mode = encoded_blend_mode ^ kBlendModeDefault;
+  paint.setBlendMode(static_cast<DlBlendMode>(blend_mode));
+
+  uint32_t style = uint_data[kStyleIndex];
+  paint.setDrawStyle(static_cast<DlDrawStyle>(style));
+
+  float stroke_width = float_data[kStrokeWidthIndex];
+  paint.setStrokeWidth(stroke_width);
+
+  float stroke_miter_limit = float_data[kStrokeMiterLimitIndex];
+  paint.setStrokeMiter(stroke_miter_limit + kStrokeMiterLimitDefault);
+
+  uint32_t stroke_cap = uint_data[kStrokeCapIndex];
+  paint.setStrokeCap(static_cast<DlStrokeCap>(stroke_cap));
+
+  uint32_t stroke_join = uint_data[kStrokeJoinIndex];
+  paint.setStrokeJoin(static_cast<DlStrokeJoin>(stroke_join));
+
+  paint.setInvertColors(uint_data[kInvertColorIndex] != 0);
+
+  paint.setDither(uint_data[kDitherIndex] != 0);
+
+  switch (uint_data[kMaskFilterIndex]) {
+    case kNull:
+      break;
+    case kBlur:
+      SkBlurStyle blur_style =
+          static_cast<SkBlurStyle>(uint_data[kMaskFilterBlurStyleIndex]);
+      double sigma = float_data[kMaskFilterSigmaIndex];
+      DlBlurMaskFilter dl_filter(blur_style, sigma);
+      if (dl_filter.skia_object()) {
+        paint.setMaskFilter(&dl_filter);
```

Comment:
This creates a copy. You could construct a shared_ptr using make_shared instead and then just allow that shared ref to be adopted into the paint.